jetty-quic/quic-common/src/main/java/org/eclipse/jetty/quic/common/QuicSession.java

```diff
@@ -0,0 +1,415 @@
+    protected QuicSession(Executor executor, Scheduler scheduler, ByteBufferPool byteBufferPool, QuicheConnection quicheConnection, QuicConnection connection, SocketAddress remoteAddress)
+    {
+        this.executor = executor;
+        this.scheduler = scheduler;
+        this.byteBufferPool = byteBufferPool;
+        this.quicheConnection = quicheConnection;
+        this.connection = connection;
+        this.flusher = new Flusher(scheduler);
+        this.strategy = new AdaptiveExecutionStrategy(new Producer(), executor);
+        this.remoteAddress = remoteAddress;
+        LifeCycle.start(strategy);
+    }
+
+    public Executor getExecutor()
+    {
+        return executor;
+    }
+
+    public Scheduler getScheduler()
+    {
+        return scheduler;
+    }
+
+    public ProtocolQuicSession getProtocolQuicSession()
+    {
+        return protocolSession;
+    }
+
+    public String getNegotiatedProtocol()
+    {
+        return quicheConnection.getNegotiatedProtocol();
+    }
+
+    public void onOpen()
+    {
+        getOrCreateStreamEndPoint(0);
+    }
+
+    public int fill(long streamId, ByteBuffer buffer) throws IOException
+    {
+        return quicheConnection.drainClearTextForStream(streamId, buffer);
+    }
+
+    public int flush(long streamId, ByteBuffer buffer) throws IOException
+    {
+        int flushed = quicheConnection.feedClearTextForStream(streamId, buffer);
+        flush();
+        return flushed;
+    }
+
+    public void flushFinished(long streamId) throws IOException
+    {
+        quicheConnection.feedFinForStream(streamId);
+        flush();
+    }
+
+    public boolean isFinished(long streamId)
+    {
+        return quicheConnection.isStreamFinished(streamId);
+    }
+
+    public void shutdownInput(long streamId) throws IOException
+    {
+        quicheConnection.shutdownStream(streamId, false);
+    }
+
+    public void shutdownOutput(long streamId) throws IOException
+    {
+        quicheConnection.shutdownStream(streamId, true);
+    }
+
+    public void onClose(long streamId)
+    {
+        endpoints.remove(streamId);
+    }
+
+    public SocketAddress getLocalAddress()
+    {
+        return connection.getEndPoint().getLocalSocketAddress();
+    }
+
+    public SocketAddress getRemoteAddress()
+    {
+        return remoteAddress;
+    }
+
+    public boolean isConnectionEstablished()
+    {
+        return quicheConnection.isConnectionEstablished();
+    }
+
+    public void setConnectionId(QuicheConnectionId quicheConnectionId)
+    {
+        this.quicheConnectionId = quicheConnectionId;
+    }
+
+    public void process(SocketAddress remoteAddress, ByteBuffer cipherBufferIn) throws IOException
+    {
+        // While the connection ID remains the same,
+        // the remote address may change so store it again.
+        this.remoteAddress = remoteAddress;
+
+        quicheConnection.feedCipherText(cipherBufferIn, remoteAddress);
+
+        if (quicheConnection.isConnectionEstablished())
```

Comment:
maybe first check for `protocolSession!=null` as we don't know if/when the established state might revert to false during a shutdown?